### PR TITLE
Add alt-shim support (fixes #3634)

### LINKED ISSUE #1606: Support shims to both console and GUI apps
When using any tools, like `sysinternals` for example, an empty command window (the shim) is open, and has to be manually closed. 

The ideal way would be that the shim itself be a "win32" app. Currently all shims are console programs. i.e, not linked as "win32" apps for gui mode.

EDIT:
Suggestion, allow for properties in the manifest to configure which binaries are gui shims, and which should be commandline shims, and use shims accordingly.

### LINKED ISSUE #1896: Handle Control-C in shim wrapper
When user hits Ctrl-C in cmd.exe session when running a program wrapped by Shim.exe, terminal session gets corrupted (hard to describe, command prompt is displayed but no commands can be started, Far Manager stops to respond, etc). The process started by Shim is not terminated

### LINKED ISSUE #2006: windows cmd gvim.exe shim waits for subprocess
I am using scoop in a windows cmd (perhaps why i am the only person seeing this). The `gvim.exe` shim waits for the subprocess to terminate, and this is not desirable. The `gvim.ps1` shim works just fine: running `gvim` opens a new windows and a new prompt is available.

I'm not sure how the `gvim.exe` shim is created but it would be nice if it did something equivalent to `start / b ..\path\to\current\gvim.exe $*`.

I'm happy to provide any additional information, though I'm not sure what may be needed or how to get it ;-). I'm sure there is a way to make `.ps1` files executable from windows cmd, but regardless `gvim.exe` from powershell exhibits the same undesired behavior.

### LINKED ISSUE #2339: python in cmd Ctrl+C problem
### process:
1. open cmd.exe 
2. enter python in cmd
3. in python REPL/shell 【press】【Ctrl+C】
4. now enter a string, will eval by cmd
5. loop with: cmd command -> python command -> cmd command

## python install from python.org don't have (Ctrl+C)KeyboardInterrupt problem

---
OS: 10.0.17134 64bit
Python: 3.7

### LINKED ISSUE #3294: Killing the shim process does not kill the child process
This issue originates from a problem with dbtest from project mgo (Mongo driver for Go): https://github.com/globalsign/mgo/issues/328. I believe this issue is a generic issue with scoop, which is exemplified with mongod+dbtest.

**Observed behaviour**
When mongod is installed using scoop, a shallow executable called a "shim" is created in the path (usually in %USERPROFILE%\scoop\shims). The actual mongod.exe lies somewhere else (normally %USERPROFILE%\scoop\apps\mongodb\current).
When you launch mongod from the command line, the shim launches the actual mongod as a child process. The same happens when dbtest launches mongod, but when it tries to stop the server only the shim gets terminated, the actual server process stays alive and is adopted by userinit.

**Expected behaviour**
Killing the shim should kill the spawned process.

### LINKED ISSUE #3634: [Proposal] Replace current C#-based shim with C version
Per:
#3294: Killing the shim process does not kill the child process,
#2921: Wait for console applications only,
#2339: python in cmd Ctrl+C problem,
#2006: Windows cmd gvim.exe shim waits for subprocess,
#1896: Handle Control-C in shim wrapper,
#1606: Support shims to both console and GUI apps,
and [extras/#2801](https://github.com/lukesampson/scoop-extras/issues/2801): Putty shims open a blank window,
I would like us to consider replacing our current C#-based [shim code,](
https://github.com/lukesampson/scoop/blob/master/supporting/shimexe/shim.cs) (or is it at [ScoopInstaller/Shim](https://github.com/ScoopInstaller/Shim) ?) with this C-based code, at https://github.com/71/scoop-better-shimexe .

Its [readme](https://github.com/71/scoop-better-shimexe/blob/master/README.md) explains the benefits. I can't think of any drawbacks.

### PR DESCRIPTION
**Pull request by rasa**
Tested and working locally with both 71 and kiennq shims. See #3634 for details.
Hopefully:
Fixes #3294: Killing the shim process does not kill the child process,
Fixes #2921: Wait for console applications only,
Fixes #2339: python in cmd Ctrl+C problem,
Fixes #2006: Windows cmd gvim.exe shim waits for subprocess,
Fixes #1896: Handle Control-C in shim wrapper,
Fixes #1606: Support shims to both console and GUI apps,
Fixes lukesampson/scoop-extras#2801 : Putty shims open a blank window,

### CODE REVIEW

**Comment by rasa** [comment]
Testing instructions are [here](https://github.com/lukesampson/scoop/issues/3634#issuecomment-633730051).

**Comment by rasa** [comment]
Can those that have [tested](https://github.com/lukesampson/scoop/issues/3634#issuecomment-633730051) this branch, please report your success/failure here. I tested both alternate shims and both worked fine.

**Comment by apommel** [comment]
I tested them too and it is definitely an improvement. I did not encounter any issues, and the console flickering when launching a GUI from Win+R is gone. Keyboard interrupts also work correctly.

**Comment by rasa** [comment]
If there are no objextions, I will merge this into develop

**Comment by chawyehsu** [comment]
@rasa I suggest releasing previous commits in develop branch into the master branch before merging this (https://github.com/lukesampson/scoop/compare/develop).

**Comment by zxcplm911** [comment]
gh pr checkout 3998

**Comment by benferse** [comment]
A little late to the party, but I've been using this for a little over a week with no issues. Keyboard interrupts work as expected everywhere, which was my biggest problem.

File: lib/core.ps1 [hunk 1]
```diff
@@ -608,6 +608,18 @@ powershell -noprofile -ex unrestricted `"& '$resolved_path' $arg %args%;exit `$l
     }
 }
 
+function get_shim_path() {
+    $shim_path = "$(versiondir 'scoop' 'current')\supporting\shimexe\bin\shim.exe"
+    $shim_version = get_config 'shim' 'default'
+    switch ($shim_version) {
+        '71' { $shim_path = "$(versiondir 'scoop' 'current')\supporting\shims\71\shim.exe" }
+        'kiennq' { $shim_path = "$(versiondir 'scoop' 'current')\supporting\shims\kiennq\shim.exe" }
+        'default' { true }
```
Comment (kiennq) [inline lib/core.ps1:617]: shouldn't this `$true` instead?
```suggestion
        'default' { $true }
```

**Comment by trajano** [comment]
Do you have instructions on how to enable this on my local scoop installation?

**Comment by alkuzad** [comment]
This is mandatory to merge to continue recommend scoop to people, what are the problems that stop it from being added to master ? Branch/Repo shenanigans are not interesting to me.

**Comment by rasa** [comment]
OK, based on the feedback above, I'd will do the following after freshening this PR with the fix above, kiennq's latest release (2.2.1), and a little more testing:
1. Merge the current develop into master
2. Merge this PR into develop
3. Merge the updated develop into master

I feel confident this is not introduce any issues, but if it does, we can always roll back, if needed.

**Comment by woacademy** [comment]
@rasa when will this make it to master?

**Comment by rasa** [comment]
OK, let's do it now. We can always roll back if there are issues.

**Comment by rasa** [comment]
OK, develop has been merged into master. I also tagged the release at 2020-11-26.

**Pull request merged**